README.md
# Genetic Maze

Can machines learn? if so, how much processing power is needed? Seeking to demonstrate the answers to these questions, I developed the Genetic Maze project. A program that uses the concept of *Genetic Algorithm* to make a machine learn to perform a specific task: *Transpose a maze*

## Getting Started

The instructions here will make the project run on your machine. For more code and development information check the documentation folders as well as the *\detailInfo.html* file

### Prerequisites

 - An average computer
 - A browser
  

### Web Implementation
You can find a online implementation [here]([http://www.grad.inf.puc-rio.br/~c1920354/GeneticMaze/main.html]) 

### Installing

The program was made entirely using pure javascript (including some graphic libraries listed below) so to run it just download the entire folder and open the file *'main.html'*
By default this file will open a page in your default browser, from there you can generate a maze (or create your own) for the cells to try to transpose.
I recommend using Google Chrome for the app.

### Changing some parameters
If you want to be able to change some features of the program, such as the cell learning rate, mutation ratio and a number of other things.
All settings are found in the *\ scripts \ config.js file*

This is an example of a parameter:

    'lineSize': 0.3,  //Size of the line dividing the cells of the grid || Default is 0.3
To change it, just change the value after the colon, in this case it would be 0.3
All parameters have their Default value and a description

### Contribuing

Contribuitons are very wellcome. I believe this project is a good opportunity to learn how genetic algorithms work. Modifying and adding new features is always a process of learning.
If you are interested in contributing please check open an Issue/PR


## Libraries used in this project

* [P5.js]([https://p5js.org/](https://p5js.org/)) - Used in the graphic part to draw various elements
* [Chart.js]([https://www.chartjs.org/](https://www.chartjs.org/)) - Using to render the charts

## Author

* **João P Maia** - *Owner, first version* 

## License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details


Documentation/changelog.txt
Author: Joao Pedro maia

*V1 - April 6, 2020
*V1.01 - May 31, 2021

scripts/main.js
// Author: Joao P Maia

//main.js
//Backbone of the entire program, takes care of drawings through p5.js, moves the population, takes care of the main functions
//Global Variables and import settings
var cols = configuration.colsNumber; // number of rows
var rows = configuration.rowsNumber; // number of cols
var end, start; // end and start points
var w, h; // width and height for each cell
var life = configuration.lifeSpan; // life is = to the size of the gene, each iteneration is 1 gene
var populationSize = configuration.populationSize; // size of the population
var currentGen = 0; // define de current gene of the entire population
var generation = 1; // current generation
var walls = 0; // number of walls
var bSolution = { 'steps': configuration.lifeSpan, 'generation': 'none' } //Starts an empty solution that in the future will be replaced by a real solution (the best solution)


//Lists
var grid = [];
var population = [];

//Game State
var start_game = false;


//
// Main Code
//


// P5 functions 
function setup() {
    var canvas = createCanvas(920, 920);
    canvas.parent("teste")

    //Height and width
    w = Math.floor(920 / cols);
    h = Math.floor(920 / rows);

    // Make a 2D array
    for (i = 0; i < cols; i++) {
        grid[i] = new Array(rows);
    }

    //Make the cells of the grid
    for (i = 0; i < cols; i++) {
        for (j = 0; j < rows; j++) {
            grid[i][j] = new Cell(i, j);
        }

    }

    // make population
    for (i = 0; i < populationSize; i++) {
        population.push(new Unit(0, 0))

    }

    // make the population genes
    for (k = 0; k < population.length; k++) {
        population[k].newGene(life);
        population[k].generation = generation;
    }

    //defines the beginning and end as the two ends of the grid and give then colors
    start = grid[0][0];
    grid[0][0].color = "#4f98ca"

    end = grid[cols - 1][rows - 1];
    grid[cols - 1][rows - 1].color = "#4f98ca"



    updatePopulation()// update DOM

}

function draw() {
    frameRate(configuration.frameRate);//How many frames per second will the simulation run | 0 to 60
    displayGrid(); // Draw the grid
    if (start_game) {

        //moves the entire population to the current gene
        if (currentGen <= life) {
            movePopulation(currentGen, population);
            currentGen++;
        } else {
            populationFitness(population); // calculate the fitness of each individual
            // update DOM
            updateScore(population) 
            updateWinner(population,end);
            updateBestSolution(population);
            // update the Chart
            chartAddAverage(generation, averageScore(population));

            population = crossover(population, generation, bSolution); // Crossover and start a new population - New generation is born
            generation++;
            updateGeneration(generation) // update DOM
            currentGen = 0;
        }
        //shows de population
        populationShow(w, h);

    }


}

//find the best solution based on number of steps
function bestSolution(population){
    var average = 0;
    var total = 0;
    for(k=0;k<population.length;k++){
        if(population[k].winner == true){
            if(population[k].steps<bSolution.steps){
                bSolution = population[k];
            }
            average += population[k].steps
            total++
        }
    }
    chartAddSteps((average/total).toFixed(2))
}
//
// End of the main
//


//
// Displays Functions
// 

// Display the grid
function displayGrid() {
    for (i = 0; i < cols; i++) {
        for (j = 0; j < rows; j++) {
            grid[i][j].display(w, h, 255);
        }
    }
}

// Display each Unit
function populationShow() {
    for (i = 0; i < population.length; i++) {
        population[i].show(w, h);
    }
}

//
// Walls Functions
//

// Place the walls
function mouseClicked() {
    var a = Math.floor(mouseX / w);
    var b = Math.floor(mouseY / h);
    if (grid[a][b].wall == true) {
        grid[a][b].wall = false;
    } else {
        if (grid[a][b] != start && grid[a][b] != end) {
            grid[a][b].wall = true;

        }
    }
    return false;
}

// place walls
function mouseDragged() {
    a = Math.floor(mouseX / w);
    b = Math.floor(mouseY / h);
    if (grid[a][b] != start && grid[a][b] != end) {
        grid[a][b].wall = true;
    }
    return false;
}

//
//Butons Functions
//

function btnStart() {
    start_game = true;
    document.getElementById("btnStart").disabled = true;
    updateDom(0,0,0,0)
}

function btnRandom() {
    for (i = 0; i < cols; i++) {
        for (j = 0; j < rows; j++) {
            r = Math.floor(random(0, 100));
            if (r <configuration.randomWallRatio*100) {
                grid[i][j].wall = true;
            }
        }
    }
    grid[start.x][start.y].wall = false;
    grid[end.x][end.y].wall = false;
}

function btnExport(){
    csvExport(generation);
}

function icnModal(){
    $('#modalInfo').modal('show');
}


scripts/unity.js
//unity.js
//Take care of each unit, this file is like the individual. Here it has all its "features", methods and functions
// import settings
var maxFit = configuration.maxFitnessDist; //the maximum fitness is calculated is the perfect distance between the beginning and the end
var maxstep = configuration.lifeSpan;
var radius = configuration.circleradius;//Unit radius

function Unit(x, y) {
    this.x = x;
    this.y = y;
    this.generation = 0;
    this.gen = [];
    this.fitness = 0;
    this.prob = 0 // probability of been chosen as a partener
    this.steps = 0;
    this.winner = false;
    this.check = false;
    this.color = "red";

    // create a new gene sequency
    this.newGene = function(size) {
        for (i = 0; i < size; i++) {
            var r = Math.floor(p5.random(0, 5));
            this.gen.push(r);
        }
    }

    this.show = function(w, h) {
        p5.fill(this.color);
        p5.circle(this.x * w + w / 2, this.y * h + h / 2, radius);
    }

    this.calculateFitness = function(xgoal, ygoal) {
        if (xgoal == this.x && ygoal == this.y) {
            fit = pow(maxFit, 2);
        } else {
            fit = pow(maxFit - manhattanDist(this.x, this.y, xgoal, ygoal), 2);
        }
        this.fitness = Math.floor(fit);
    }

    this.checkWin = function(xgoal, ygoal) {
        if (this.x == xgoal && this.y == ygoal) {
            this.winner = true;
            this.color = "green";
        }
    }

}

//The taxicab metric is also known as Manhattan distance 
function manhattanDist(x1, y1, x2, y2) {
    return Math.abs(x1 - x2) + Math.abs(y1 - y2);
}

scripts/genetic.js
//genetic.js
//Takes care of all genetics: crossover and mutations
// import settings
var populationSize = configuration.populationSize;
var mutationRate = configuration.mutationRate;
var life = configuration.lifeSpan;


//crossover
function crossover(pop, generetion,bsSolution) {

    var newPopulation = [];

    //Normalize the distances/fitness and calculate the probability of each organism being chosen as a partner for the next generation
    var sum = 0;

    for (i = 0; i < pop.length; i++) {
        sum += pop[i].fitness;
    }

    for (i = 0; i < pop.length; i++) {
        pop[i].prob = (pop[i].fitness / sum); // probability of each organism being chosen, 0 to 1
    }


    // Crossover: gather half the genes of each organism chosen as a partner
    while (newPopulation.length <= populationSize) {
        var partnerA = findPartner(pop);
        var partnerB = findPartner(pop);


        var child = new Unit(0, 0);

        var cut = Math.floor(random(1, life));

        for (i = 0; i < cut; i++) {
            child.gen[i] = partnerA[i];
        }
        for (i = cut; i < life; i++) {
            child.gen[i] = partnerB[i];
        }

        child.generation = generetion + 1;


        //Calculates the probability of each child mutating, if it occurs, the entire gene is modified
        var r = Math.floor(p5.random(0, 101))
        if (r < mutationRate * 100) {
            child.gen = shuffleMutation(child.gen);    
            child.color = "purple";
        }


        newPopulation.push(child);

    }
    //Transfer the best solution to the population
    if (bsSolution.generation > 0) {
        var bestChild = new Unit(0, 0);
        bestChild.color = ' yellow';
        bestChild.gen = bsSolution.gen;
        bestChild.generation = bsSolution.generation
        newPopulation.push(bestChild);
    }

    return newPopulation;
}



//Function to choose the partner based on the probability assigned at the start of the crossover
function findPartner(list) {
    var index = 0;
    var r = Math.random();

    while (r > 0) {
        r = r - list[index].prob;
        index++;
    }
    index--;
    return list[index].gen;
}



//The taxicab metric is also known as Manhattan distance 
function manhattanDist(x1, y1, x2, y2) {
    return Math.abs(x1 - x2) + Math.abs(y1 - y2);
}



//Some mutation types

//shuffleMutation
function shuffleMutation(array) {
    var currentIndex = array.length, temporaryValue, randomIndex;

    // While there remain elements to shuffle...
    while (0 !== currentIndex) {
  
      // Pick a remaining element...
      randomIndex = Math.floor(Math.random() * currentIndex);
      currentIndex -= 1;
  
      // And swap it with the current element.
      temporaryValue = array[currentIndex];
      array[currentIndex] = array[randomIndex];
      array[randomIndex] = temporaryValue;
    }
  
    return array;
  }
  
//RandomMutation (not used)
function randomMutation(array){
    for (i = 0; i < array.length; i++) {
        array[i] = Math.floor(p5.random(0, 5));
    }
    return child
}

// bit flip mutation (not used)
function bitflipMutation(array){
    for(k=0;k<=configuration.mutationSize*100;k++){
        rdngen = Math.floor(Math.random(0,array.length))
        array[rdngen] = Math.floor(Math.random(0,5))
    }
    return array
}

scripts/graphic.js
//graphic.js
//Responsible for plotting the data on the chart as well as exporting the data

var ctx = document.getElementById('myChart').getContext('2d');

//Create the chart
var myChart = new Chart(ctx, {
    type: 'line',
    data: {
        
        labels: [],
        datasets: [{
            label: 'Winners',
            borderColor:'#FFBE57',
            data: [],
            borderWidth: 2,
            fill: false
        },
        {
            label: 'Average fitness',
            borderColor:'#A53DFF',
            data: [],
            borderWidth: 2,
            fill: false
        },
        {
            label: 'Average number of steps',
            borderColor:'#22B357',
            data: [],
            borderWidth: 2,
            fill: false
        }]
    },
    options: {
        scales: {
            xAxes: [{
                scaleLabel: {
                  display: true,
                  labelString: 'Generations'
                }}],

            yAxes: [{
                ticks: {
                    beginAtZero: true
                }
            }]
        }
    }
});

//Add fitness average to the graph
function chartAddAverage(gen,num){
    myChart.data.labels.push(gen);
    myChart.data.datasets[1].data.push(num)
    myChart.update();
}
//Add winner average to the graph
function chartAddWinners(num){
    myChart.data.datasets[0].data.push(num)
    myChart.update();
}
//Add steps average to the graph (winners only)
function chartAddSteps(num){
    myChart.data.datasets[2].data.push(num)
    myChart.update();
}

//Exports all chart data in .csv format
function csvExport(generations){
     
    var csv = 'sep=,\nGeneration,Winners ,Average Fitness, Average Steps\n';
 
    for(k=0;k<generations-1;k++){
        csv = csv +  k + ","
        csv = csv +  myChart.data.datasets[0].data[k] + ","
        csv = csv +  myChart.data.datasets[1].data[k] + ","
        csv = csv +  myChart.data.datasets[2].data[k] + "\n"
    };
  
    var hiddenElement = document.createElement('a');
    hiddenElement.href = 'data:text/csv;charset=utf-8,' + encodeURI(csv);
    hiddenElement.target = '_blank';
    hiddenElement.download = 'geneticMaze.csv';
    hiddenElement.click()};

scripts/movement.js
//movement.js
// Convert each item of the gene into motion, check the walls, the winners and calculate fitness

//Global varibles
var totalWinner = 0;//Number of generation winners


//Move the population || 1 rigth, 2 left, 3 botton, 4 down
function movePopulation(gen, population) {
    var k = gen;
    for (i = 0; i < population.length; i++) {

        move = population[i].gen[k]; // current step

        population[i].checkWin(end.x, end.y) // checks if the current cell/unit is at the end

        if (population[i].winner == true && population[i].check == false) {
            population[i].check = true;
            totalWinner++;
            updateTotalWinners(totalWinner)
        }

        if (checkMove(population[i], k) && population[i].winner == false) {
            if (move == 1 && population[i].x + 1 >= 0 && population[i].x + 1 <= cols) {
                population[i].x = population[i].x + 1;
                population[i].steps++
            }
            if (move == 2 && population[i].x - 1 >= 0 && population[i].x - 1 <= cols) {
                population[i].x = population[i].x - 1;
                population[i].steps++
            }
            if (move == 3 && population[i].y + 1 >= 0 && population[i].y + 1 <= rows) {
                population[i].y = population[i].y + 1;
                population[i].steps++
            }
            if (move == 4 && population[i].y - 1 >= 0 && population[i].y - 1 <= rows) {
                population[i].y = population[i].y - 1;
                population[i].steps++
            }

        }
    }
}

//Calculate the fitness for each individual
function populationFitness(population) {
    for (i = 0; i < population.length; i++) {
        population[i].calculateFitness(end.x, end.y);
    }
}

//Calculates if the next move is possible, that is, there are no walls and it is within the grid boundary
function checkMove(u, i) {
    var nextStep = u.gen[i];
    var uX = u.x;
    var uY = u.y;
    if (nextStep == 1 && uX + 1 < cols && uX + 1 >= 0) {
        if (grid[uX + 1][uY].wall) {
            return false;
        } else {
            return true;
        }
    }
    if (nextStep == 2 && uX + 1 < cols && uX - 1 >= 0) {
        if (grid[uX - 1][uY].wall) {
            return false;
        } else {
            return true;
        }
    }
    if (nextStep == 3 && uY + 1 < rows && uY + 1 >= 0) {
        if (grid[uX][uY + 1].wall) {
            return false;
        } else {
            return true;
        }
    }
    if (nextStep == 4 && uY - 1 < rows && uY - 1 >= 0) {
        if (grid[uX][uY - 1].wall) {
            return false;
        } else {
            return true;
        }
    }
}


//The taxicab metric is also known as Manhattan distance 
function manhattanDist(x1, y1, x2, y2) {
    return Math.abs(x1 - x2) + Math.abs(y1 - y2);
}

scripts/domManipulation.js
//domManipulation.js
//This file manipulates the DOM, updates the winners, the best score, etc.

//Best solution found, when none is found the value of the generation is 'none'. The best solution is found by the smallest number of steps among the winners
var bSolution = { 'steps': configuration.lifeSpan, 'generation': 'none' }

//Displays population, gene and mutation information found in the config.js file
function updatePopulation(){
    document.getElementById("size").innerText = "Size: " + configuration.populationSize;
    document.getElementById("lifespan").innerText = "Lifespan: " + configuration.lifeSpan;
    document.getElementById("mutationRate").innerText = "Mutation Rate: " + configuration.mutationRate;
}

//Receive the generation and show it on the screen above the canvas
function updateGeneration(generation){
    document.getElementById("generation").innerText = generation;
}
//Receive the best score and show in the side menu
function updateScore(population){
    bScore = bestScore(population)
    document.getElementById("bestScore").innerText = "Best Score: " + bScore;
}
//Receive the population find the winners and show in the side menu
function updateWinner(population,end,generation){
    var winners = 0
    for(k=0;k<population.length;k++){
        if(population[k].x == end.x && population[k].y == end.y){
            winners++
        }
    }
    document.getElementById("winners").innerText = "Last Generation winners: " + winners;
    chartAddWinners(winners)
}
//Receives the winners of the generation
function updateTotalWinners(totalWinners){
    document.getElementById("totalWinners").innerText = "Total Winners: " + totalWinners;
}

//Update the best solution
function updateBestSolution(population){
    bestSolution(population)
    document.getElementById("bestSolution").innerText = "Best solution found in: Generation " + bSolution.generation
}




//Calculates the best score
function bestScore(population){
    var best = 0;
    for(k=0;k<population.length;k++){
        if(population[k].fitness>best){
            best = population[k].fitness;
        }
    }
    return Math.sqrt(best);
}
// Calculates the Average Score
function averageScore(population){
    var scores = 0;
    for(k=0;k<population.length;k++){
        scores = scores + Math.sqrt(population[k].fitness)
    }
    return (scores/population.length).toFixed(2);
}
// Calculates the best Solution | Best solution is the cell that took the least amount of steps among the winners|Calculates the average of steps
function bestSolution(population){
    var average = 0;
    var total = 0;
    for(k=0;k<population.length;k++){
        if(population[k].winner == true){
            if(population[k].steps<bSolution.steps){
                bSolution = population[k];
            }
            average += population[k].steps
            total++
        }
    }
    chartAddSteps((average/total).toFixed(2))
    return bSolution
}


scripts/config.js
//Config.js
//This file contains all program settings
//Please, for all percetages use values beetwen 0 and 1
var configuration = {
    'author' :'Joao P Maia',
    'version' :'v1,April 2020',
    //GRID
    'frameRate': 60,//How many frames per second will the simulation run | 0 to 60
    'rowsNumber': 40, // Define the number of columns and rows || Sefault: 40x40
    'colsNumber': 40,//Same as above
    'lineSize': 0.1, //Size of the line dividing the cells of the grid || Default is 0.3
    'randomWallRatio': 0.20, // When selected, the "random" button defines the percentage of cell that will be transformed into walls || Deafult: 0.20

    //POPULATION
    'circleradius': 10, //Unit size | default 10
    'populationSize': 90, //Population Size || Default: 90

    // GENETICS //
    'lifeSpan': 360, // Gene size, number of instructions that each Unity/Organism will perform || Default: 360 || To modify values, use the formula: (rows * cols) * 9/40 = lifespan
    'maxFitnessDist': manhattanDist(0, 0, 39, 39),//Definition for the Fitness function, look at the documentation
    'mutationRate': 0.1, //Mutation rate at each crossover || Default: 0.1

    //MUTATIONS
    'mutationType': 0, //Defines the type of mutation to be applied | default 0
    'mutationSize': 0.20 //Percentage of the gene that is mutated during the mutation || Default .20
}

//Calculates the distance
function manhattanDist(x1,y1,x2,y2){
    return Math.abs(x1-x2) + Math.abs(y1-y2);
}